### PR TITLE
Custom web3 provider

### PR DESCRIPTION
**Pull request by SergeyZhukovsky**
Resolves https://github.com/brave/brave-browser/issues/14823

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by SergeyZhukovsky** [comment]
post-init failed due to audit problems that are not related to the current PR

File: browser/brave_content_browser_client.cc [hunk 1]
```diff
@@ -236,6 +241,26 @@ void BindCosmeticFiltersResources(
       std::move(receiver));
 }
 
+#if BUILDFLAG(BRAVE_WALLET_ENABLED)
+void BindBraveWalletProvider(
+    content::RenderFrameHost* const frame_host,
+    mojo::PendingReceiver<brave_wallet::mojom::BraveWalletProvider> receiver) {
+  auto* web_contents = content::WebContents::FromRenderFrameHost(frame_host);
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc:248]: you can get the BrowserContext directly from the RenderFrameHost

Comment (bridiver) [reply browser/brave_content_browser_client.cc:248]: also it would be better for `BraveWalletServiceFactory::GetInstance()->GetForProfile` to return nullptr for IsTor so the logic is consistent no matter where it gets called

File: browser/brave_content_browser_client.cc [hunk 2]
```diff
@@ -236,6 +241,26 @@ void BindCosmeticFiltersResources(
       std::move(receiver));
 }
 
+#if BUILDFLAG(BRAVE_WALLET_ENABLED)
+void BindBraveWalletProvider(
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc:245]: should be `MaybeBindBraveWalletProvider`

File: browser/brave_content_browser_client.cc (hunk 1 above)
Comment (bridiver) [reply browser/brave_content_browser_client.cc:248]: cc @bbondy ^^ because the existing code gets the regular profile for Tor so we need null checks in other places it gets used as well

File: browser/brave_content_browser_client.cc [hunk 3]
```diff
@@ -271,6 +296,16 @@ void BraveContentBrowserClient::BrowserURLHandlerCreated(
   ChromeContentBrowserClient::BrowserURLHandlerCreated(handler);
 }
 
+void BraveContentBrowserClient::RenderProcessWillLaunch(
+    content::RenderProcessHost* host) {
+  Profile* profile = Profile::FromBrowserContext(host->GetBrowserContext());
+  Profile* original_profile = profile->GetOriginalProfile();
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc:302]: not sure why upstream does it like this, but it should be 
```content::BrowserContext* BraveRendererUpdaterFactory::GetBrowserContextToUse(
    content::BrowserContext* context) const {
  return chrome::GetBrowserContextRedirectedInIncognito(context);
}```

File: browser/profiles/brave_renderer_updater.cc [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  mojo::AssociatedRemote<brave::mojom::BraveRendererConfiguration>
+      renderer_configuration;
+  channel->GetRemoteAssociatedInterface(&renderer_configuration);
+
+  return renderer_configuration;
+}
+
+void BraveRendererUpdater::OnPrimaryAccountChanged(
+    const signin::PrimaryAccountChangeEvent& event) {
+  if (event.GetEventTypeFor(signin::ConsentLevel::kSync) ==
```
Comment (bridiver) [inline browser/profiles/brave_renderer_updater.cc:87]: this actually isn't relevant for us, we can remove all the IdentityManager related code

File: browser/brave_content_browser_client.cc [hunk 5]
```diff
@@ -286,6 +321,14 @@ void BraveContentBrowserClient::RegisterBrowserInterfaceBindersForFrame(
       render_frame_host, map);
   map->Add<cosmetic_filters::mojom::CosmeticFiltersResources>(
       base::BindRepeating(&BindCosmeticFiltersResources));
+
+#if BUILDFLAG(BRAVE_WALLET_ENABLED)
+  if (base::FeatureList::IsEnabled(
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc:326]: `brave_wallet::IsNativeWalletEnabled`

File: browser/profiles/brave_renderer_updater.cc [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/profiles/brave_renderer_updater.h"
+
+#include <utility>
+
+#include "base/bind.h"
+#include "brave/common/brave_renderer_configuration.mojom.h"
+#include "brave/components/brave_wallet/brave_wallet_constants.h"
+#include "brave/components/brave_wallet/pref_names.h"
```
Comment (bridiver) [inline browser/profiles/brave_renderer_updater.cc:13]: this needs to be in a buildflag for brave wallet enabled

File: browser/profiles/brave_renderer_updater.cc [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    UpdateRenderer(&renderer_configuration);
+}
+
+void BraveRendererUpdater::UpdateRenderer(
+    mojo::AssociatedRemote<brave::mojom::BraveRendererConfiguration>*
+        renderer_configuration) {
+  bool use_brave_web3_provider = (static_cast<BraveWalletWeb3ProviderTypes>(
+                                      brave_wallet_web3_provider_.GetValue()) ==
+                                  BraveWalletWeb3ProviderTypes::BRAVE_WALLET) &&
+                                 !is_tor_profile_;
```
Comment (bridiver) [inline browser/profiles/brave_renderer_updater.cc:106]: BraveRendererUpdater shouldn't have to know that this is disabled for tor profiles specifically. Maybe `brave_wallet::IsAllowedForProfile(Profile* profile)` which could also be used in `BraveWalletServiceFactory::GetForProfile` so they are always in sync

File: browser/ui/BUILD.gn [hunk 8]
```diff
@@ -459,12 +459,16 @@ source_set("ui") {
...
+  if (brave_wallet_enabled) {
+    if (!is_android) {
+      sources += [
+        "webui/brave_wallet_ui.cc",
+        "webui/brave_wallet_ui.h",
+      ]
+      deps += [ "//brave/components/brave_wallet_ui:generated_resources" ]
+    }
+
+    deps += [ "//brave/components/brave_wallet/resources:generated_resources" ]
```
Comment (bridiver) [inline browser/ui/BUILD.gn:471]: why is this here?

File: components/brave_wallet/renderer/DEPS [hunk 9]
```diff
@@ -0,0 +1,8 @@
+include_rules = [
+  "+brave/common",
```
Comment (bridiver) [inline components/brave_wallet/renderer/DEPS:2]: brave/common is not allowed in components. You need to move `BraveWalletRenderFrameObserver` into brave/renderer/brave_wallet to avoid this layering violation

File: components/brave_wallet/browser/BUILD.gn [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import("//build/config/features.gni")
```
Comment (bridiver) [inline components/brave_wallet/browser/BUILD.gn:1]: adding components/brave_wallet/browser is a problem because it is already a browser-only component as indicated by having all the classes in the root of components/brave_wallet. We can't just add `browser` without moving everything that is currently in the root of components/brave_wallet into components/brave_wallet/browser

File: components/brave_wallet/common/BUILD.gn [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import("//mojo/public/tools/bindings/mojom.gni")
```
Comment (bridiver) [inline components/brave_wallet/common/BUILD.gn:1]: same as brave_wallet/browser. A component must either not have common/browser/renderer/etc... or only have them. It can't have files in both the root and brower/common/etc...

File: components/brave_wallet/browser/brave_wallet_provider.cc [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+#include "brave/components/brave_wallet/browser/brave_wallet_provider.h"
+
+#include <utility>
+
+#include "brave/components/brave_wallet/brave_wallet_service.h"
+
+namespace brave_wallet {
+
+BraveWalletProvider::BraveWalletProvider(BraveWalletService* wallet_service)
```
Comment (bridiver) [inline components/brave_wallet/browser/brave_wallet_provider.cc:14]: This class seems like an unnecessary extra layer. I think it would be simpler for BraveWalletService to just implement `brave_wallet::mojom::BraveWalletProvider` directly

File: components/brave_wallet/browser/brave_wallet_provider.cc [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    : wallet_service_(wallet_service), weak_factory_(this) {}
+
+BraveWalletProvider::~BraveWalletProvider() {}
+
+void BraveWalletProvider::Request(const std::string& json_payload,
+                                  RequestCallback callback) {
+  if (!wallet_service_)
```
Comment (bridiver) [inline components/brave_wallet/browser/brave_wallet_provider.cc:21]: this should never be null here because we shouldn't be creating `BraveWalletProvider` in the first place if it is

File: browser/brave_content_browser_client.cc [hunk 14]
```diff
@@ -236,6 +241,26 @@ void BindCosmeticFiltersResources(
...
+    return;
+
+  auto* profile =
+      Profile::FromBrowserContext(web_contents->GetBrowserContext());
+  BraveWalletService* service =
+      BraveWalletServiceFactory::GetInstance()->GetForProfile(
+          Profile::FromBrowserContext(profile));
+
+  mojo::MakeSelfOwnedReceiver(
+      std::make_unique<brave_wallet::BraveWalletProvider>(service),
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc:259]: I don't think it's safe to pass the service here like this because I think it's possible that an ipc message could come in after the service has been destroyed. I think you need to pass in a WeakPtr ref here

Comment (bridiver) [reply browser/brave_content_browser_client.cc:259]: wdyt @iefremov ? The renderer processes should all shut down before the profile (and service) is destroyed, but could there still be an unprocessed message in the queue?

File: components/brave_wallet/browser/brave_wallet_provider.h [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+#include <string>
+
+#include "base/memory/weak_ptr.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+
+class BraveWalletService;
+
+namespace brave_wallet {
+
+class BraveWalletProvider final
```
Comment (bridiver) [inline components/brave_wallet/browser/brave_wallet_provider.h:20]: this should be `BraveWalletProviderImpl` to distinguish it from the interface. Also the general convention for naming these seems to be *Host to BraveWalletHost?

File: components/brave_wallet/resources/brave_wallet_provider.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+(function() {
+    if (!window.ethereum) {
+      return
+    }
+
+    // TODO: fill window.ethereum object with functions
```
Comment (bridiver) [inline components/brave_wallet/resources/brave_wallet_provider.js:11]: ?

Comment (bridiver) [reply components/brave_wallet/resources/brave_wallet_provider.js:11]: we're defining this in c++, why do we need to execute any JS at all?

File: components/brave_wallet/renderer/brave_wallet_js_handler.cc [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+#include "gin/function_template.h"
+#include "third_party/blink/public/common/browser_interface_broker_proxy.h"
+#include "third_party/blink/public/web/blink.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+#include "third_party/blink/public/web/web_script_source.h"
+#include "ui/base/resource/resource_bundle.h"
+
+namespace {
+
+static base::NoDestructor<std::string> g_provider_script("");
```
Comment (bridiver) [inline components/brave_wallet/renderer/brave_wallet_js_handler.cc:24]: same as the other comment, why do we need a script for anything?

File: components/brave_wallet/renderer/brave_wallet_js_handler.cc [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+  v8::HandleScope handle_scope(isolate);
+  if (context.IsEmpty())
+    return;
+
+  v8::Context::Scope context_scope(context);
+
+  CreateWorkerObject(isolate, context);
+}
+
+void BraveWalletJSHandler::CreateWorkerObject(v8::Isolate* isolate,
```
Comment (bridiver) [inline components/brave_wallet/renderer/brave_wallet_js_handler.cc:69]: This method name doesn't really make sense because this doesn't have anything to do with workers. CreateEthereumObject?

File: components/brave_wallet/renderer/brave_wallet_render_frame_observer.cc [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    return;
+
+  native_javascript_handle_->AddJavaScriptObjectToFrame(context);
+}
+
+void BraveWalletRenderFrameObserver::DidCreateNewDocument() {
+  if (!dynamic_params_.brave_use_native_wallet || !native_javascript_handle_)
+    return;
+
+  native_javascript_handle_->InjectScript();
```
Comment (bridiver) [inline components/brave_wallet/renderer/brave_wallet_render_frame_observer.cc:41]: if we keep this injected script it needs to go in `DidCreateScriptContext` so it gets created for each isolated world

Comment (bridiver) [reply components/brave_wallet/renderer/brave_wallet_render_frame_observer.cc:41]: also doesn't this run before `DidCreateScriptContext` which would mean window.ethereum would always be null for the script

File: browser/profiles/brave_renderer_updater.cc (hunk 4 above)
Comment (SergeyZhukovsky) [reply browser/profiles/brave_renderer_updater.cc:87]: I did it without IdentityManager, but afterwards added as it updates renderers after accounts switch, so perhaps it makes sense.

File: components/brave_wallet/renderer/brave_wallet_render_frame_observer.cc (hunk 19 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/renderer/brave_wallet_render_frame_observer.cc:41]: `DidCreateScriptContext` runs before or on script injection. If you debug it you will see on a callstack that `ExecuteScript` triggers `DidCreateScriptContext` in the same thread if it has not been called before that

File: browser/brave_content_browser_client.cc (hunk 14 above)
Comment (SergeyZhukovsky) [reply browser/brave_content_browser_client.cc:259]: Made is as a WeakPtr to be safe.

File: components/brave_wallet/browser/brave_wallet_provider.cc (hunk 13 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/browser/brave_wallet_provider.cc:22]: the check should stay as I remade it to have a WeakPtr in case of IPC message comes but service is already dead

File: browser/ui/BUILD.gn (hunk 8 above)
Comment (SergeyZhukovsky) [reply browser/ui/BUILD.gn:471]: removed it

File: components/brave_wallet/browser/BUILD.gn (hunk 10 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/browser/BUILD.gn:3]: everything has been moved to browser level

File: components/brave_wallet/common/BUILD.gn (hunk 11 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/common/BUILD.gn:2]: everything has been moved to browser folder

File: components/brave_wallet/resources/brave_wallet_provider.js (hunk 16 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/resources/brave_wallet_provider.js:11]: removed scripts execution

File: components/brave_wallet/renderer/brave_wallet_js_handler.cc (hunk 17 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/renderer/brave_wallet_js_handler.cc:24]: removed script execution

File: components/brave_wallet/renderer/brave_wallet_render_frame_observer.cc (hunk 19 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/renderer/brave_wallet_render_frame_observer.cc:41]: removed script execution

File: browser/profiles/brave_renderer_updater.cc (hunk 7 above)
Comment (SergeyZhukovsky) [reply browser/profiles/brave_renderer_updater.cc:113]: done

File: browser/brave_content_browser_client.cc (hunk 1 above)
Comment (SergeyZhukovsky) [reply browser/brave_content_browser_client.cc]: made a check via `brave_wallet::IsAllowedForProfile(Profile* profile)`

File: browser/brave_content_browser_client.cc (hunk 3 above)
Comment (SergeyZhukovsky) [reply browser/brave_content_browser_client.cc:302]: ok, done

File: components/brave_wallet/renderer/brave_wallet_js_handler.cc (hunk 18 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/renderer/brave_wallet_js_handler.cc:69]: renamed

File: components/brave_wallet/browser/brave_wallet_provider.h (hunk 15 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/browser/brave_wallet_provider.h:20]: renamed

File: components/brave_wallet/renderer/DEPS (hunk 9 above)
Comment (SergeyZhukovsky) [reply components/brave_wallet/renderer/DEPS:2]: done

**Comment by SergeyZhukovsky** [comment]
post-init failed because of the audit problem that is not related to that PR

File: browser/brave_content_browser_client.cc [hunk 20]
```diff
@@ -271,6 +298,17 @@ void BraveContentBrowserClient::BrowserURLHandlerCreated(
   ChromeContentBrowserClient::BrowserURLHandlerCreated(handler);
 }
 
+void BraveContentBrowserClient::RenderProcessWillLaunch(
+    content::RenderProcessHost* host) {
+  Profile* profile = Profile::FromBrowserContext(
+      chrome::GetBrowserContextRedirectedInIncognito(
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc:304]: this needs to go in `BraveRendererUpdaterFactory::GetBrowserContextToUse` so all calls to `BraveRendererUpdaterFactory::GetForProfile` return the same thing

File: browser/profiles/brave_renderer_updater.cc (hunk 4 above)
Comment (bridiver) [reply browser/profiles/brave_renderer_updater.cc:92]: We will never have anything that will change based on google login because that is all upstream by definition. We have no control over google login servers so we can't store any new information there.

File: renderer/sources.gni [hunk 21]
```diff
@@ -3,12 +3,26 @@
...
+import("//brave/components/brave_wallet/common/buildflags/buildflags.gni")
+
 brave_chrome_renderer_sources = [
   "//brave/renderer/brave_content_renderer_client.cc",
   "//brave/renderer/brave_content_renderer_client.h",
+  "//brave/renderer/brave_render_thread_observer.cc",
+  "//brave/renderer/brave_render_thread_observer.h",
 ]
 
 brave_chrome_renderer_public_deps = [
```
Comment (bridiver) [inline renderer/sources.gni:15]: I think you're missing a brave wallet buildflag dep here

File: browser/profiles/brave_renderer_updater.cc [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+void BraveRendererUpdater::UpdateRenderer(
+    mojo::AssociatedRemote<brave::mojom::BraveRendererConfiguration>*
+        renderer_configuration) {
+#if BUILDFLAG(BRAVE_WALLET_ENABLED)
+  bool use_brave_web3_provider = (static_cast<BraveWalletWeb3ProviderTypes>(
+                                      brave_wallet_web3_provider_.GetValue()) ==
+                                  BraveWalletWeb3ProviderTypes::BRAVE_WALLET) &&
+                                 brave_wallet::IsAllowedForProfile(profile_);
+
+  (*renderer_configuration)
```
Comment (bridiver) [inline browser/profiles/brave_renderer_updater.cc:114]: nit: it doesn't matter right now, but it would be better to use the buildflag to set use_brave_web3_provider = false instead of wrapping this whole block because this will have to change when we add other properties

**Review by bridiver — APPROVED** [review]
approved based on making the few quick fixes left

File: browser/brave_content_browser_client.cc (hunk 20 above)
Comment (SergeyZhukovsky) [reply browser/brave_content_browser_client.cc]: done

File: browser/profiles/brave_renderer_updater.cc (hunk 22 above)
Comment (SergeyZhukovsky) [reply browser/profiles/brave_renderer_updater.cc:96]: yeah, I thought about that, but in any case that place has to be changed when new pref come.

**Comment by SergeyZhukovsky** [comment]
post-init failed due to unrelated audit

**Review by bbondy — APPROVED** [review]
